pkg/asset/installconfig/aws/validation.go

```diff
@@ -0,0 +1,153 @@
+	allErrs := field.ErrorList{}
+	if config.Platform.AWS != nil {
+		allErrs = append(allErrs, validatePlatform(ctx, meta, field.NewPath("platform", "aws"), config.Platform.AWS, config.Networking, config.Publish)...)
+
+		if config.ControlPlane != nil && config.ControlPlane.Platform.AWS != nil {
+			allErrs = append(allErrs, validateMachinePool(ctx, meta, field.NewPath("controlPlane", "platform", "aws"), config.Platform.AWS, config.ControlPlane.Platform.AWS)...)
+		}
+		if len(config.Compute) > 0 {
+			fldPath := field.NewPath("compute")
+			for idx, compute := range config.Compute {
+				fldPath := fldPath.Index(idx)
+				if compute.Platform.AWS != nil {
+					allErrs = append(allErrs, validateMachinePool(ctx, meta, fldPath.Child("platform", "aws"), config.Platform.AWS, compute.Platform.AWS)...)
+				}
+			}
+		}
+	}
+	return allErrs.ToAggregate()
+}
+
+func validatePlatform(ctx context.Context, meta *Metadata, fldPath *field.Path, platform *awstypes.Platform, networking *types.Networking, publish types.PublishingStrategy) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if len(platform.Subnets) > 0 {
+		allErrs = append(allErrs, validateSubnets(ctx, meta, fldPath.Child("subnets"), platform.Subnets, networking, publish)...)
+	}
+	return allErrs
+}
+
+func validateSubnets(ctx context.Context, meta *Metadata, fldPath *field.Path, subnets []string, networking *types.Networking, publish types.PublishingStrategy) field.ErrorList {
+	allErrs := field.ErrorList{}
+	privateSubnets, err := meta.PrivateSubnets(ctx)
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, subnets, err.Error()))
+	}
+	privateSubnetsIdx := map[string]int{}
+	for idx, id := range subnets {
+		if _, ok := privateSubnets[id]; ok {
+			privateSubnetsIdx[id] = idx
+		}
+	}
+	if len(privateSubnets) == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, subnets, "No private subnets found"))
+	}
+
+	publicSubnets, err := meta.PublicSubnets(ctx)
+	if err != nil {
+		return append(allErrs, field.Invalid(fldPath, subnets, err.Error()))
+	}
+	publicSubnetsIdx := map[string]int{}
+	for idx, id := range subnets {
+		if _, ok := publicSubnets[id]; ok {
+			publicSubnetsIdx[id] = idx
+		}
+	}
+
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, privateSubnets, privateSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateSubnetCIDR(fldPath, publicSubnets, publicSubnetsIdx, networking.MachineCIDR)...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, privateSubnets, privateSubnetsIdx, "private")...)
+	allErrs = append(allErrs, validateDuplicateSubnetZones(fldPath, publicSubnets, publicSubnetsIdx, "public")...)
+
+	privateZones := sets.NewString()
+	publicZones := sets.NewString()
+	for _, subnet := range privateSubnets {
+		privateZones.Insert(subnet.Zone)
+	}
+	for _, subnet := range publicSubnets {
+		publicZones.Insert(subnet.Zone)
+	}
+	if publish != types.InternalPublishingStrategy && !publicZones.IsSuperset(privateZones) {
```

Comment:
Since we call this validation stuff after `defaults.SetInstallConfigDefaults` in `InstallConfig.finish`, can we use `publish == ExternalPublishingStrategy && !publicZones.IsSuperset(privateZones)` to drop a `!=` here?

Reply:
@wking fixed in https://github.com/openshift/installer/compare/a96182f3f4d7c7b69bd674bd918dd72a532f6ce4..9ce0bb39f5cbc2d581705794c8e16933cae350d4